lib/handoff/transfer_list.c

```diff
@@ -0,0 +1,407 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (c) 2023, Linaro Limited
+ */
+
+#include <assert.h>
+#include <common/debug.h>
+#include <inttypes.h>
+#include <lib/transfer_list.h>
+#include <lib/utils_def.h>
+#include <string.h>
+
+#define ADD_OVERFLOW(a, b, res) __builtin_add_overflow((a), (b), (res))
+
+#define SUB_OVERFLOW(a, b, res) __builtin_sub_overflow((a), (b), (res))
+
+#define MUL_OVERFLOW(a, b, res) __builtin_mul_overflow((a), (b), (res))
+
+#define ROUNDUP_OVERFLOW(v, size, res) (__extension__({ \
+	typeof(*(res)) __roundup_tmp = 0; \
+	typeof(v) __roundup_mask = (typeof(v))(size) - 1; \
+	\
+	ADD_OVERFLOW((v), __roundup_mask, &__roundup_tmp) ? 1 : \
+		(void)(*(res) = __roundup_tmp & ~__roundup_mask), 0; \
+}))
+
+#define ALIGN_MASK(p) ((1 << (p)) - 1)
+
+#define IS_ALIGNED(x, a) (((x) & ALIGN_MASK((typeof(x))(a))) == 0)
+
+#define ALIGN(x, a) (__extension__({ \
+	typeof(x) __a = (a); \
+	\
+	((x) + ALIGN_MASK((__a))) & ~ALIGN_MASK((__a)); \
+}))
+
+/*******************************************************************************
+ * Creating a transfer list in a reserved memory region specified
+ * Compliant to 2.4.5 of Firmware handoff specification (v0.9)
+ * Return pointer to the created transfer list or NULL on error
+ ******************************************************************************/
+struct transfer_list_header *transfer_list_init(void *p, size_t max_size)
+{
+	struct transfer_list_header *tl = p;
+
+	if (!IS_ALIGNED((uintptr_t)p, TRANSFER_LIST_INIT_MAX_ALIGN) ||
+	    !IS_ALIGNED(max_size, TRANSFER_LIST_INIT_MAX_ALIGN) ||
+	    max_size < sizeof(*tl))
+		return NULL;
+
+	memset(tl, 0, max_size);
+	tl->signature = TRANSFER_LIST_SIGNATURE;
+	tl->version = TRANSFER_LIST_VERSION;
+	tl->hdr_size = sizeof(*tl);
+	tl->alignment = TRANSFER_LIST_INIT_MAX_ALIGN; // initial max align
+	tl->size = sizeof(*tl); // initial size is the size of header
+	tl->max_size = max_size;
+	transfer_list_update_checksum(tl);
+	return tl;
+}
+
+/*******************************************************************************
+ * Relocating a transfer list to a reserved memory region specified
+ * Compliant to 2.4.6 of Firmware handoff specification (v0.9)
+ * Return true on success or false on error
+ ******************************************************************************/
+bool transfer_list_relocate(struct transfer_list_header *tl, void *p, size_t max_size)
+{
+	uintptr_t new_tl_base = 0;
+	uintptr_t *np = (uintptr_t *)p;
+
+	if (!tl || !p || max_size == 0)
+		return false;
+
+	new_tl_base = ALIGN(*np, tl->alignment);
+
+	if (new_tl_base < *np)
+		new_tl_base += (1 << tl->alignment);
+
+	if (new_tl_base - *np + tl->size > max_size)
+		return false;
+
+	memmove((void *)new_tl_base, tl, tl->size);
+	((struct transfer_list_header *)new_tl_base)->max_size = max_size - (new_tl_base - *np);
+	*np = new_tl_base;
+	transfer_list_update_checksum((struct transfer_list_header *)(*np));
```

Comment:
I think you need to rework the whole concept here.  np and all the weird uintptr_t casts don't make much sense.  Instead define a new_tl ptr that you fill in normally and hand over to this function

Reply:
Function `transfer_list_relocate` is refactored into a better shape, now the arg 'addr' represents a physical address, and the return of the function is the new TL handle. Some duplicated logics inside the function are removed as well.